File: packages/zynth-automation/android/src/main/java/dev/zynth/automation/ZynthAutomationModule.kt
package dev.zynth.automation

import android.content.pm.ApplicationInfo
import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthRuntime
import com.zynth.kit.runtime.ZynthSyncModule
import com.zynth.kit.runtime.ZynthArgs
import org.json.JSONObject

internal class ZynthAutomationModule(
    private val runtime: ZynthRuntime,
) : ZynthModule, ZynthSyncModule {

    override val name: String = "Automation"
    private var productionInspectionEnabled: Boolean = false

    override val exportedMethods: List<String> = listOf("read", "configure")
    override val protectedMethods: List<String> = listOf("read", "configure")

    override fun call(method: String, args: ZynthArgs): JSONObject {
        return when (method) {
            "read" -> JSONObject().put("result", runtime.getUIManager().snapshot(normalizedReadOptions(args)))
            "configure" -> JSONObject().put("result", JSONObject(configure(args)))
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }

    override fun callSync(method: String, args: ZynthArgs): Any? {
        return when (method) {
            "read" -> runtime.getUIManager().snapshot(normalizedReadOptions(args))
            "configure" -> configure(args)
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }

    private fun parseOptions(args: ZynthArgs): JSONObject {
        return try {
            JSONObject(args.getMapAt(0))
        } catch (e: Exception) {
            JSONObject()
        }
    }

    private fun configure(args: ZynthArgs): Map<String, Any> {
        val options = parseOptions(args)
        productionInspectionEnabled = options.optBoolean("enableProductionInspection", false)
        return mapOf("productionInspectionEnabled" to productionInspectionEnabled)
    }

    private fun normalizedReadOptions(args: ZynthArgs): JSONObject {
        val options = parseOptions(args)
        if (!isDebugBuild() && !productionInspectionEnabled) {
            // Heavy fields are disabled in production by default to reduce overhead.
            options.put("includeResolvedStyles", false)
            options.put("includeComponentState", false)
            options.put("includeText", false)
        }
        return options
    }

    private fun isDebugBuild(): Boolean {
        val flags = runtime.root.context.applicationInfo.flags
        return (flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
    }
}

File: packages/zynth-automation/ios/ZynthAutomationModule.swift
import Foundation
import ZynthKit

final class ZynthAutomationModule: NSObject, ZynthModule, ZynthSyncModule {
  let name: String = "Automation"

  var exportedMethods: [String] {
    return ["read", "configure"]
  }

  var protectedMethods: [String] {
    return ["read", "configure"]
  }

  private weak var runtime: ZynthRuntime?
  private var productionInspectionEnabled = false

  init(runtime: ZynthRuntime) {
    self.runtime = runtime
    super.init()
  }

  func call(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "read":
      return ["result": try readSnapshot(args: args)]
    case "configure":
      return ["result": configure(args: args)]
    default:
      throw ZynthModuleError.methodNotExported(module: name, method: method)
    }
  }

  func callSync(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "read":
      return try readSnapshot(args: args)
    case "configure":
      return configure(args: args)
    default:
      throw ZynthModuleError.methodNotExported(module: name, method: method)
    }
  }

  private func readSnapshot(args: ZynthArgs) throws -> [String: Any] {
    guard let runtime else {
      throw ZynthModuleError.runtimeDeallocated
    }
    var options = (try? args.asDict()) ?? [:]
    if !isDebugBuild && !productionInspectionEnabled {
      // Heavy fields are disabled in production by default to reduce overhead.
      options["includeResolvedStyles"] = false
      options["includeComponentState"] = false
      options["includeText"] = false
    }
    return runtime.uiManager.snapshot(options) as? [String: Any] ?? [:]
  }

  private func configure(args: ZynthArgs) -> [String: Any] {
    let config = try? args.asDict()
    productionInspectionEnabled = config?["enableProductionInspection"] as? Bool ?? false
    return ["productionInspectionEnabled": productionInspectionEnabled]
  }

  private var isDebugBuild: Bool {
#if DEBUG
    return true
#else
    return false
#endif
  }
}
